text/0000-task-builder.md

```diff
@@ -0,0 +1,200 @@
+- Feature Name: Async Builder
+- Start Date: 2020-05-31
+- RFC PR: (leave this empty)
+- Neon Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+The `Task` API was introduced in Neon `v0.4.0` and has proven to be incredibly valuable to developers. It empowers users to execute long running tasks on the libuv thread pool without blocking the main javascript thread.
+
+The `TaskBuilder` API expands upon this concept with an improved ownership model and ergonomics.
+
+# Motivation
+[motivation]: #motivation
+
+The current `Task` trait passes `self` as a shared reference (`&self`) to `fn perform`. Multiple users (https://github.com/neon-bindings/neon/pull/283) (https://github.com/neon-bindings/neon/pull/520) have found this ownership limiting. Many usages require a mutable or owned `self`. Work-arounds require error prone runtime checks (e.g., `Cell` and `Option`).
+
+While unlikely to cause issues, changing `&self` to `&mut self` is a breaking change. Users could have manually called `fn perform` in a context that relied on shared references.
+
+Instead of introducing a breaking change, a new API is introduced with an _owned_ `self`. The following addtional changes are made:
+
+* The new API relies on closures instead of a `trait`; thanks to many improvements in the borrow checker and closure ergonomics
+* The new API returns a single `Output` generic rather than separate `Output` and `Error` types. This simplifies infallible tasks.
```

Comment:
Mention that it's still more expressive, e.g.: "This simplifies infallible tasks and is still expressive enough to allow users to use a `Result` type to represent fallible tasks."